lib/ansible/plugins/action/package.py

```diff
@@ -77,31 +78,57 @@ def run(self, tmp=None, task_vars=None):
                     except KeyError:
                         raise AnsibleActionFail('Could not detect a package manager. Try using the "use" option.')
 
-            if module and module != 'auto':
-                if not self._shared_loader_obj.module_loader.has_plugin(module):
-                    raise AnsibleActionFail('Could not find a matching action for the "%s" package manager.' % module)
-                else:
-                    # run the 'package' module
-                    new_module_args = self._task.args.copy()
-                    if 'use' in new_module_args:
-                        del new_module_args['use']
-
+            if action and action != 'auto':
+                module_context = None
+
+                # see if we use custom mapped, use orig if not
+                action = C.PACKAGE_MANAGERS.get(action, action)
+                # prefix with ansible.legacy to eliminate external collisions while still allowing library/ override
+                if action in self.BUILTIN_PKG_MGR_MODULES:
+                    action = f'ansible.legacy.{action}'
+
+                # find what to execute, action plugins having priority
+                has_action_plugin = self._shared_loader_obj.module_loader.has_plugin(action)
+                if not has_action_plugin:
+                    module_context = self._shared_loader_obj.module_loader.find_plugin_with_context(action, collection_list=self._task.collections)
+                    if module_context and module_context.resolved and module_context.action_plugin:
+                        # module itself specifies action plugin
+                        action = module_context.action_plugin
+                        has_action_plugin = True
+
+                # prep to run he action
+                new_module_args = self._task.args.copy()
+                if 'use' in new_module_args:
+                    del new_module_args['use']
+
+                if has_action_plugin:
+                    display.vvvv("Chose {action!r} action plugin")
+                    new_task = new_task = self._task.copy()
+                    new_task.args.update(new_module_args)
+                    pkg_action = self._shared_loader_obj.action_loader.get(action,
+                                                                           task=new_task,
+                                                                           connection=self._connection,
+                                                                           play_context=self._play_context,
+                                                                           loader=self._loader,
+                                                                           templar=self._templar,
+                                                                           shared_loader_obj=self._shared_loader_obj)
+                    display.vvvv("Running {action!r}")
+                    result.update(pkg_action.run(task_vars=task_vars))
```

Comment:
Looks like `pkg_action` is sometimes `None`.

Reply:
we check via `has_action_plugin` .. not sure how it gets there being none, unless it fails on instantiation .. but can add check there